sys/xtimer/xtimer.c

```diff
@@ -54,82 +54,88 @@ void _xtimer_tsleep(uint32_t offset, uint32_t long_offset)
-     */
-    /*
-     * Note: last_wakeup _must never_ specify a time in the future after
-     * _xtimer_periodic_sleep returns.
-     * If this happens, last_wakeup may specify a time in the future when the
-     * next call to _xtimer_periodic_sleep is made, which in turn will trigger
-     * the overflow logic above and make the next timer fire too early, causing
-     * last_wakeup to point even further into the future, leading to a chain
-     * reaction.
-     *
-     * tl;dr Don't return too early!
-     */
-    uint32_t offset = target - now;
-    DEBUG("xps, now: %9" PRIu32 ", tgt: %9" PRIu32 ", off: %9" PRIu32 "\n", now, target, offset);
-    if (offset < XTIMER_PERIODIC_SPIN) {
-        _xtimer_spin(offset);
-    }
-    else {
-        if (offset < XTIMER_PERIODIC_RELATIVE) {
-            /* NB: This will overshoot the target by the amount of time it took
-             * to get here from the beginning of xtimer_periodic_wakeup()
-             *
-             * Since interrupts are normally enabled inside this function, this time may
-             * be undeterministic. */
-            target = _xtimer_now() + offset;
+        /*
+        * For large offsets, set an absolute target time.
+        * As that might cause an underflow, for small offsets, set a relative
+        * target time.
+        * For very small offsets, spin.
+        */
+        /*
+        * Note: last_wakeup _must never_ specify a time in the future after
+        * _xtimer_periodic_sleep returns.
+        * If this happens, last_wakeup may specify a time in the future when the
+        * next call to _xtimer_periodic_sleep is made, which in turn will trigger
+        * the overflow logic above and make the next timer fire too early, causing
+        * last_wakeup to point even further into the future, leading to a chain
+        * reaction.
+        *
+        * tl;dr Don't return too early!
+        */
+        uint32_t offset = target - now;
+        DEBUG("xps, now: %9" PRIu32 ", tgt: %9" PRIu32 ", off: %9" PRIu32 "\n", now, target, offset);
+        if (offset < XTIMER_PERIODIC_SPIN) {
+            _xtimer_spin(offset);
+            timer->callback(timer->arg);
         }
-        mutex_lock(&mutex);
-        DEBUG("xps, abs: %" PRIu32 "\n", target);
-        _xtimer_set_absolute(&timer, target);
-        mutex_lock(&mutex);
-    }
-out:
+        else {
+            if (offset < XTIMER_PERIODIC_RELATIVE) {
+                /* NB: This will overshoot the target by the amount of time it took
+                * to get here from the beginning of xtimer_periodic_wakeup()
+                *
+                * Since interrupts are normally enabled inside this function, this time may
+                * be undeterministic. */
+                target = _xtimer_now() + offset;
+            }
+            DEBUG("xps, abs: %" PRIu32 "\n", target);
+            _xtimer_set_absolute(timer, target);
+        }
+    } while (0);
     *last_wakeup = target;
```

Comment:
this is a problem for the _msg variant, the last wakeup time will increment before that time has actually passed, leading to the situation described above in the long comment about not returning early.
Calling xtimer_periodic_msg twice without waiting for the first message to send will mess up the timer.